### PR TITLE
feat: Add rocm_smi input to monitor AMD GPUs

### LINKED ISSUE #9601: Add plugin to monitor AMD GPUs using ROCm System Management Interface rocm-smi
## Feature Request

As it is done with nvdia GPUs would be nice to have the possibility to query metrics from AMD devices using the `rocm-smi` [0] utility.

### Proposal:

Write a plugin that parses the JSON output provided by `rocm-smi` executable and extracts metrics to be sent via Telegraf. 

### Current behavior:

 Currently I could not find any effort in that direction.

### Use case: 

Very convenient in all those scenario where ROCm [1] is deployed as computing environment for GPGPU applications.

### Some pointers:
[0] https://rocmdocs.amd.com/en/latest/index.html
[1] https://rocmdocs.amd.com/en/latest/Installation_Guide/ROC-smi.html

### PR DESCRIPTION
**Pull request by mconcas**
### Required for all PRs:

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary) (e.g. feat: or fix:)

resolves #9601

Added an input plugin to gather AMD GPUs metrics using ROCm System Management Interface rocm-smi.

### CODE REVIEW

**Comment by mconcas** [comment]
Hello, any feedback on this?

**Review by Hipska — CHANGES_REQUESTED** [review]
So basically, this is just `inputs.exec` + `parsers.json` into one? I don't see the need to add this plugin, while the same can already be achieved by setting some configurations.

File: etc/telegraf.conf [hunk 1]
```diff
@@ -315,6 +315,30 @@
...
+#   ## Type of metrics grouping used when pushing to Azure Data Explorer.
+#   ## Default is "TablePerMetric" for one table per different metric.
+#   ## For more information, please check the plugin README.
+#   # metrics_grouping_type = "TablePerMetric"
+#
+#   ## Name of the single table to store all the metrics (Only needed if metrics_grouping_type is "SingleTable").
+#   # table_name = ""
+#
+
+
```
[the remark this answers is not in the document]
Comment (mconcas) [reply etc/telegraf.conf:341]: Thanks for spotting, I'll remove it.

**Comment by mconcas** [comment]
> So basically, this is just `inputs.exec` + `parsers.json` into one? I don't see the need to add this plugin, while the same can already be achieved by setting some configurations.

As a general comment: the approach I chose was heavily inspired to the [nvidia-smi plugin](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/nvidia_smi) that was already present in Telegraf, so I wnt straight with that, I admittedly was not aware of the possibility to combine plugins so I can't tell if the combination of plugins you mention is enough to fullfill the case.
A part from the parsing there is indeed some data digestion to produce indirect metrics and tag/fields names I'll have a look into it.

File: plugins/inputs/rocm_smi/README.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
...CI Bus": "0000:04:00.0", "VRAM Total Memory (B)": "17163091968", "VRAM Total Used Memory (B)": "17776640", "VIS_VRAM Total Memory (B)": "268435456", "VIS_VRAM Total Used Memory (B)": "13557760", "GTT Total Memory (B)": "17163091968", "GTT Total Used Memory (B)": "25608192", "ASD firmware version": "553648152", "CE firmware version": "79", "DMCU firmware version": "0", "MC firmware version": "0", "ME firmware version": "163", "MEC firmware version": "432", "MEC2 firmware version": "432", "PFP firmware version": "186", "RLC firmware version": "93", "RLC SRLC firmware version": "0", "RLC SRLG firmware version": "0", "RLC SRLS firmware version": "0", "SDMA firmware version": "430", "SDMA2 firmware version": "430", "SMC firmware version": "00.28.54.00", "SOS firmware version": "0x0008015d", "TA RAS firmware version": "00.00.00.00", "TA XGMI firmware version": "00.00.00.00", "UVD firmware version": "0x422b1100", "VCE firmware version": "0x39060400", "VCN firmware version": "0x00000000", "Card model": "0xc1e", "Card vendor": "Advanced Micro Devices, Inc. [AMD/ATI]", "Card SKU": "D05101", "(Topology) Numa Node": "0", "(Topology) Numa Affinity": "0"}, "system": {"Driver version": "5.9.25"}}
```
[the remark this answers is not in the document]
Comment (mconcas) [reply plugins/inputs/rocm_smi/README.md:50]: Ok, thanks. This was really mimicked from the `nvidia-smi` readme file.

[the remark this answers is not in the document]
Comment (reimda) [reply plugins/inputs/rocm_smi/README.md:50]: I don't mind adding a plugin that just runs a program and parses the json if it's commonly used. Without a plugin for this, each user needs to go through the config process, including learning to use the exec and json plugins. AMD GPUs are very common so a dedicated plugin could be useful.

**Comment by reimda** [comment]
> > So basically, this is just `inputs.exec` + `parsers.json` into one? I don't see the need to add this plugin, while the same can already be achieved by setting some configurations.
> 
> As a general comment: the approach I chose was heavily inspired to the [nvidia-smi plugin](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/nvidia_smi) that was already present in Telegraf, so I went straight with that, I admittedly was not aware of the possibility to combine plugins so I can't tell if the combination of plugins you mention is enough to fullfill the case.
> A part from the parsing there is indeed some data digestion to produce indirect metrics and tag/fields names, in future iteration this can become more complex. I'll have a look into it.

I don't mind adding a plugin that just runs a program and parses the json if it's commonly used. Without a plugin for this, each user needs to go through the config process, including learning to use the exec and json plugins and coming up with a good metric format. AMD GPUs are very common so a dedicated plugin could be useful.

If you can quickly make the changes Hipska requested I think we can get this into the 1.20.0 release candidate which comes out Sept 1 (tomorrow).

File: plugins/inputs/all/all.go [hunk 3]
```diff
@@ -159,6 +159,7 @@ import (
 	_ "github.com/influxdata/telegraf/plugins/inputs/rethinkdb"
 	_ "github.com/influxdata/telegraf/plugins/inputs/riak"
 	_ "github.com/influxdata/telegraf/plugins/inputs/riemann_listener"
+	_ "github.com/influxdata/telegraf/plugins/inputs/rocm_smi"
```
Comment (reimda) [inline plugins/inputs/all/all.go:162]: I understand "rocm_smi" is the name of the program that gathers data from the GPU, but I don't think it's a name that people will recognize. Could we change the plugin name to include AMD GPU or Radeon instead?

**Review by reimda — CHANGES_REQUESTED** [review]

File: plugins/inputs/all/all.go (hunk 3 above)
Comment (mconcas) [reply plugins/inputs/all/all.go]: changed in `amd_rocm_smi`, can it work?

Comment (reimda) [reply plugins/inputs/all/all.go]: In their docs, AMD seems use "AMD ROCm" and "ROCm" interchangeably, so "amd_rocm_smi" seems like a good name for the plugin.

Comment (mconcas) [reply plugins/inputs/all/all.go]: great, so I renamed.

**Review by Hipska — CHANGES_REQUESTED** [review]
Could you still include an example metric in influx format please? (It could be with data from one of the tests if you like)

**Review by Hipska — APPROVED** [review]

**Review by reimda — APPROVED** [review]

**Pull request merged**